### PR TITLE
add prometheus for monitor

### PR DESCRIPTION
**Pull request by ChenQShmily**
add perf_counter_enable_prometheus, prometheus_host,prometheus_port to config.ini
perf_counter_enable_prometheus : whether to use prometheus to report data
prometheus_host : the host that pushgateway running
prometheus_port : the port pushgateway use

add _enable_prometheus, _prometheus_host, _prometheus_port to pegasus_counter_reporter.h, receive the data from config.ini

pegasus_counter_reporter.cpp
s_split(): split metrics_names that look like ***:1_0_p999, add app_id, partition_count and percent to metrics_labels SO there can have more classification methods to monitor

update()
 if(_enable_prometheus){
line 253-259
prometheus don't accept the format cs.name use, so use the temporary variable metrics_name to hold the right name; change such as "collector*app.pegasus*app.stat.multi_put_qps@1.0_p999" to "collector_app_pegasus_app_stat_multi_put_qps:1_0_p999"

line 261-268
split metrics_name like "collector_app_pegasus_app_stat_multi_put_qps:1_0_p999", then get "app_id = 1","partition_count = 0","percent = p999" as metrics_labels

line 269-278
create metrics that prometheus support to hold data, add data_labels and set data_value

line 280-281
push the metrics to pushgateway
pushgateway can push data to prometheus automatically
}

### CODE REVIEW

File: src/redis_protocol/proxy/config.ini [hunk 1]
```diff
@@ -118,11 +118,15 @@ perf_counter_cluster_name = onebox
 perf_counter_update_interval_seconds = 10
 perf_counter_enable_logging = true
 perf_counter_enable_falcon = false
+perf_counter_enable_prometheus = true
```
Comment (hycdong) [inline src/redis_protocol/proxy/config.ini]: set default prometheus is false
perf_counter_enable_prometheus = false

File: src/server/config-server.ini [hunk 2]
```diff
@@ -265,15 +265,19 @@ manual_compact_min_interval_seconds = 30
 
 perf_counter_cluster_name = onebox
 perf_counter_update_interval_seconds = 10
-perf_counter_enable_logging = false
+perf_counter_enable_logging = true
 perf_counter_enable_falcon = false
+perf_counter_enable_prometheus = true
```
Comment (hycdong) [inline src/server/config-server.ini]: perf_counter_enable_logging = false
perf_counter_enable_prometheus = false

File: src/reporter/pegasus_counter_reporter.cpp [hunk 3]
```diff
@@ -112,11 +164,17 @@ void pegasus_counter_reporter::start()
         "pegasus.server", "perf_counter_enable_logging", true, "perf_counter_enable_logging");
     _enable_falcon = dsn_config_get_value_bool(
         "pegasus.server", "perf_counter_enable_falcon", false, "perf_counter_enable_falcon");
+    _enable_prometheus = dsn_config_get_value_bool(
+        "pegasus.server", "perf_counter_enable_prometheus", true, "perf_counter_enable_prometheus");
```
Comment (hycdong) [inline src/reporter/pegasus_counter_reporter.cpp]: dsn_config_get_value_bool("pegasus.server", "perf_counter_enable_prometheus", false, "perf_counter_enable_prometheus")

File: src/reporter/pegasus_counter_reporter.cpp [hunk 4]
```diff
@@ -167,21 +225,63 @@ void pegasus_counter_reporter::update()
...
+
+        using namespace prometheus;
+        const auto labels = Gateway::GetInstanceLabel(GetHostName());
+        Gateway gateway{_prometheus_host, std::to_string(_prometheus_port), "pegasus", labels};
+        auto registry = std::make_shared<Registry>();
+
+        perf_counters::instance().iterate_snapshot(
+            [&registry, this](const dsn::perf_counters::counter_snapshot &cs) {
+                std::string metrics_name = cs.name;
+
```
Comment (hycdong) [inline src/reporter/pegasus_counter_reporter.cpp]: add some comments here to explain why you replace characters

File: src/reporter/pegasus_counter_reporter.cpp [hunk 5]
```diff
@@ -167,21 +225,63 @@ void pegasus_counter_reporter::update()
...
+        perf_counters::instance().iterate_snapshot(
+            [&registry, this](const dsn::perf_counters::counter_snapshot &cs) {
+                std::string metrics_name = cs.name;
+                
+                replace(metrics_name.begin(),metrics_name.end(),'@',':');
+                replace(metrics_name.begin(),metrics_name.end(),'.','_');
+                replace(metrics_name.begin(),metrics_name.end(),'*','_');
+                replace(metrics_name.begin(),metrics_name.end(),'(','_');
+                replace(metrics_name.begin(),metrics_name.end(),')','_');
+                ddebug("%s", cs.name.c_str());
```
Comment (hycdong) [inline src/reporter/pegasus_counter_reporter.cpp]: if(_enable_logging){
   // print metrics names
}
We do not need print metrics when not enable logging

File: src/reporter/pegasus_counter_reporter.cpp [hunk 6]
```diff
@@ -167,21 +225,63 @@ void pegasus_counter_reporter::update()
...
+                }
+                auto& gauge_family_all = BuildGauge()
+                                        .Name(metrics_name)
+                                        .Labels(
+                                            {{"service","pegasus"},{"cluster",_cluster_name},{"job",_app_name},{"port",std::to_string(_local_port)}})
+                                        .Register(*registry);
+                auto& second_gauge_all = gauge_family_all.Add({{"app_id",app[0]},{"partition_count",app[1]},{"percent",app[2]}}
+                    );
+
+                second_gauge_all.Set(cs.value);
```
Comment (hycdong) [inline src/reporter/pegasus_counter_reporter.cpp]: I could not figure out what code from line262 to line278 are doing, you can add comments or just explain in pr description.

File: src/reporter/pegasus_counter_reporter.cpp [hunk 7]
```diff
@@ -18,8 +18,50 @@
...
+
+#include <prometheus/registry.h>
+#include <prometheus/gateway.h>
+
 using namespace ::dsn;
 
+static std::string GetHostName() {
+  char hostname[1024];
+
+  if (::gethostname(hostname, sizeof(hostname))) {
```
[the remark this answers is not in the document]
Comment (ChenQShmily) [reply src/reporter/pegasus_counter_reporter.cpp]: 这个函数是prometheus-cpp给的sample，所以我就直接copy过来了，这种方法会有什么不好的地方吗

File: src/reporter/pegasus_counter_reporter.cpp [hunk 8]
```diff
@@ -43,12 +85,22 @@ pegasus_counter_reporter::pegasus_counter_reporter()
       _last_report_time_ms(0),
       _enable_logging(false),
       _enable_falcon(false),
-      _falcon_port(0)
+      _enable_prometheus(false),
+      _falcon_port(0),
```
[the remark this answers is not in the document]
Comment (ChenQShmily) [reply src/reporter/pegasus_counter_reporter.cpp]: _falcon_port(0)是之前就有的默认值，我也不清楚是谁写的，就没敢动

**Pull request closed without merging**